New compile warning errors with gcc 8.2

All builds with this version are now failing with

```
FAILED: core/CMakeFiles/drdecode.dir/string.c.o
/usr/bin/cc  -I/usr/local/google/home/hgreving/dynamorio/src/core/drlibc -I/usr/local/google/home/hgreving/dynamorio/src/core/arch/x86 -I/usr/local/google/home/hgreving/dynamorio/src/core/unix -I/usr/local/google/home/hgreving/dynamorio/src/core/arch -I/usr/local/google/home/hgreving/dynamorio/src/core/lib -I. -Iinclude/annotations -m32  -fno-strict-aliasing -fno-stack-protector -mpreferred-stack-boundary=2 -fvisibility=internal  -std=gnu99 -g3 -fno-omit-frame-pointer -fno-builtin-strcmp -Wall -Werror -Wwrite-strings -Wno-unused-but-set-variable     -DNOT_DYNAMORIO_CORE_PROPER -DSTANDALONE_DECODER -fPIC -MD -MT core/CMakeFiles/drdecode.dir/string.c.o -MF core/CMakeFiles/drdecode.dir/string.c.o.d -o core/CMakeFiles/drdecode.dir/string.c.o   -c /usr/local/google/home/hgreving/dynamorio/src/core/string.c                               
/usr/local/google/home/hgreving/dynamorio/src/core/string.c:176:1: error: ‘__memmove_chk’ alias between functions of incompatible types ‘void *(void *, const void *, size_t,  size_t)’ {aka ‘void *(void *, const
void *, unsigned int,  unsigned int)’} and ‘void *(void *, const void *, size_t)’ {aka ‘void *(void *, const void *, unsigned int)’} [-Werror=attribute-alias]                                                    
 __memmove_chk(void *dst, const void *src, size_t n, size_t dst_len)
 ^~~~~~~~~~~~~
/usr/local/google/home/hgreving/dynamorio/src/core/string.c:155:1: note: aliased declaration here
 d_r_memmove(void *dst, const void *src, size_t n)
 ^~~~~~~~~~~
/usr/local/google/home/hgreving/dynamorio/src/core/string.c:200:1: error: ‘__strncat_chk’ alias between functions of incompatible types ‘void *(char *, const char *, size_t,  size_t)’ {aka ‘void *(char *, const
char *, unsigned int,  unsigned int)’} and ‘char *(char *, const char *, size_t)’ {aka ‘char *(char *, const char *, unsigned int)’} [-Werror=attribute-alias]                                                    
 __strncat_chk(char *dst, const char *src, size_t n, size_t dst_len)
 ^~~~~~~~~~~~~
/usr/local/google/home/hgreving/dynamorio/src/core/string.c:134:1: note: aliased declaration here
 d_r_strncat(char *dest, const char *src, size_t n)
 ^~~~~~~~~~~
/usr/local/google/home/hgreving/dynamorio/src/core/string.c:188:1: error: ‘__strncpy_chk’ alias between functions of incompatible types ‘void *(char *, const char *, size_t,  size_t)’ {aka ‘void *(char *, const
char *, unsigned int,  unsigned int)’} and ‘char *(char *, const char *, size_t)’ {aka ‘char *(char *, const char *, unsigned int)’} [-Werror=attribute-alias]                                                    
 __strncpy_chk(char *dst, const char *src, size_t n, size_t dst_len)
 ^~~~~~~~~~~~~
/usr/local/google/home/hgreving/dynamorio/src/core/string.c:121:1: note: aliased declaration here
 d_r_strncpy(char *dst, const char *src, size_t n)
 ^~~~~~~~~~~
cc1: all warnings being treated as errors
[24/1227] Building C object core/CMakeFiles/dynamorio_static_nohide.dir/fragment.c.o
ninja: build stopped: subcommand failed.

```

```
hgreving@hendrik-desktop:~/dynamorio/build$ gcc --version
gcc (Debian 8.2.0-14+build1) 8.2.0
```

I have not tested whether this affect only the gcc version I am working with or any >= 8.2.

The same compiler switch also turned on warning errors about potential '\0' string termination bugs, caused by the warnings -Wstringop-truncation, -Wformat-truncation, and  -Wstringop-overflow:

```
n file included from /usr/local/google/home/hgreving/dynamorio/src/libutil/dr_config.c:42:
/usr/local/google/home/hgreving/dynamorio/src/libutil/dr_config.c: In function ‘read_options’:
/usr/local/google/home/hgreving/dynamorio/src/libutil/dr_frontend.h:105:22: error: ‘strncpy’ output truncated before terminating nul copying as many bytes from a string as its length [-Werror=stringop-truncation]
 #    define _tcsncpy strncpy
/usr/local/google/home/hgreving/dynamorio/src/libutil/dr_config.c:753:5: note: in expansion of macro ‘_tcsncpy’
     _tcsncpy(tmp, ptr, len);
     ^~~~~~~~
In file included from /usr/local/google/home/hgreving/dynamorio/src/libutil/share.h:68,
                 from /usr/local/google/home/hgreving/dynamorio/src/libutil/utils.h:37,
                 from /usr/local/google/home/hgreving/dynamorio/src/libutil/dr_config.c:37:
/usr/local/google/home/hgreving/dynamorio/src/libutil/dr_frontend.h:101:21: note: length computed here
 #    define _tcslen strlen
/usr/local/google/home/hgreving/dynamorio/src/core/lib/globals_shared.h:432:32: note: in definition of macro ‘MIN’
 #    define MIN(x, y) ((x) <= (y) ? (x) : (y))
                                ^
/usr/local/google/home/hgreving/dynamorio/src/libutil/dr_config.c:752:42: note: in expansion of macro ‘_tcslen’
     len = MIN(DR_MAX_OPTIONS_LENGTH - 1, _tcslen(ptr));
                                          ^~~~~~~
In file included from /usr/local/google/home/hgreving/dynamorio/src/libutil/dr_config.c:42:
In function ‘add_extra_option.part.2’,
    inlined from ‘add_extra_option’ at /usr/local/google/home/hgreving/dynamorio/src/libutil/dr_config.c:265:1,
    inlined from ‘read_options’ at /usr/local/google/home/hgreving/dynamorio/src/libutil/dr_config.c:860:17:
/usr/local/google/home/hgreving/dynamorio/src/libutil/dr_frontend.h:105:22: error: ‘strncpy’ output truncated before terminating nul copying as many bytes from a string as its length [-Werror=stringop-truncation]
 #    define _tcsncpy strncpy
/usr/local/google/home/hgreving/dynamorio/src/libutil/dr_config.c:277:9: note: in expansion of macro ‘_tcsncpy’
         _tcsncpy(opt_info->extra_opts[idx], opt, len);
         ^~~~~~~~
In file included from /usr/local/google/home/hgreving/dynamorio/src/libutil/share.h:68,
                 from /usr/local/google/home/hgreving/dynamorio/src/libutil/utils.h:37,
                 from /usr/local/google/home/hgreving/dynamorio/src/libutil/dr_config.c:37:
/usr/local/google/home/hgreving/dynamorio/src/libutil/dr_config.c: In function ‘read_options’:
/usr/local/google/home/hgreving/dynamorio/src/libutil/dr_frontend.h:101:21: note: length computed here
 #    define _tcslen strlen
/usr/local/google/home/hgreving/dynamorio/src/core/lib/globals_shared.h:432:32: note: in definition of macro ‘MIN’
 #    define MIN(x, y) ((x) <= (y) ? (x) : (y))
                                ^
/usr/local/google/home/hgreving/dynamorio/src/libutil/dr_config.c:274:46: note: in expansion of macro ‘_tcslen’
         len = MIN(DR_MAX_OPTIONS_LENGTH - 1, _tcslen(opt));
                                              ^~~~~~~
In file included from /usr/local/google/home/hgreving/dynamorio/src/libutil/dr_config.c:42:
In function ‘add_extra_option.part.2’,
    inlined from ‘add_extra_option’ at /usr/local/google/home/hgreving/dynamorio/src/libutil/dr_config.c:265:1,
    inlined from ‘add_extra_option_char’ at /usr/local/google/home/hgreving/dynamorio/src/libutil/dr_config.c:293:16,
    inlined from ‘dr_register_process’ at /usr/local/google/home/hgreving/dynamorio/src/libutil/dr_config.c:1170:5:
/usr/local/google/home/hgreving/dynamorio/src/libutil/dr_frontend.h:105:22: error: ‘strncpy’ output truncated before terminating nul copying as many bytes from a string as its length [-Werror=stringop-truncation]
 #    define _tcsncpy strncpy
/usr/local/google/home/hgreving/dynamorio/src/libutil/dr_config.c:277:9: note: in expansion of macro ‘_tcsncpy’
         _tcsncpy(opt_info->extra_opts[idx], opt, len);
         ^~~~~~~~
In file included from /usr/local/google/home/hgreving/dynamorio/src/libutil/share.h:68,
                 from /usr/local/google/home/hgreving/dynamorio/src/libutil/utils.h:37,
                 from /usr/local/google/home/hgreving/dynamorio/src/libutil/dr_config.c:37:
/usr/local/google/home/hgreving/dynamorio/src/libutil/dr_config.c: In function ‘dr_register_process’:
/usr/local/google/home/hgreving/dynamorio/src/libutil/dr_frontend.h:101:21: note: length computed here
 #    define _tcslen strlen
/usr/local/google/home/hgreving/dynamorio/src/core/lib/globals_shared.h:432:32: note: in definition of macro ‘MIN’
 #    define MIN(x, y) ((x) <= (y) ? (x) : (y))
                                ^
/usr/local/google/home/hgreving/dynamorio/src/libutil/dr_config.c:274:46: note: in expansion of macro ‘_tcslen’
         len = MIN(DR_MAX_OPTIONS_LENGTH - 1, _tcslen(opt));
                                              ^~~~~~~
/usr/local/google/home/hgreving/dynamorio/src/libutil/dr_config.c: In function ‘dr_unregister_process’:
/usr/local/google/home/hgreving/dynamorio/src/libutil/dr_config.c:476:39: error: ‘snprintf’ output may be truncated before the last format character [-Werror=format-truncation=]
     _snprintf(fname, fname_len, "%s/%s", dir, subdir);
                                       ^
In file included from /usr/local/google/home/hgreving/dynamorio/src/libutil/dr_config.c:42:
/usr/local/google/home/hgreving/dynamorio/src/libutil/dr_frontend.h:108:23: note: ‘snprintf’ output between 2 and 513 bytes into a destination of size 512
 #    define _snprintf snprintf
/usr/local/google/home/hgreving/dynamorio/src/libutil/dr_config.c:476:5: note: in expansion of macro ‘_snprintf’
     _snprintf(fname, fname_len, "%s/%s", dir, subdir);

```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -754,6 +754,23 @@ if (UNIX)
     if (nounused_avail)
       set(WARN "${WARN} -Wno-unused-but-set-variable")
     endif (nounused_avail)
+    # XXX i#3793: DynamoRIO manages '\0' termination and error states itself in
+    # too many places. In order to activate this warning, this code needs to get
+    # re-factored for no good reason.
+    CHECK_C_COMPILER_FLAG("-Wstringop-truncation" stringop_truncation_avail)
+    if (stringop_truncation_avail)
+      set(WARN "${WARN} -Wno-stringop-truncation")
+    endif (stringop_truncation_avail)
+    # XXX i#3793: see comment above.
+    CHECK_C_COMPILER_FLAG("-Wformat-truncation" format_truncation_avail)
+    if (format_truncation_avail)
+      set(WARN "${WARN} -Wno-format-truncation")
+    endif (format_truncation_avail)
+    # XXX i#3793: see comment above.
+    CHECK_C_COMPILER_FLAG("-Wstringop-overflow" stringop_overflow_avail)
+    if (stringop_overflow_avail)
+      set(WARN "${WARN} -Wno-stringop-overflow")
+    endif (stringop_overflow_avail)
     CHECK_C_COMPILER_FLAG("-Wtype-limits" HAVE_TYPELIMITS_CONTROL)
   else (NOT CMAKE_COMPILER_IS_CLANG)
     # clang turns off color when it's writing to a pipe, but the user may still

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -754,19 +754,19 @@ if (UNIX)
     if (nounused_avail)
       set(WARN "${WARN} -Wno-unused-but-set-variable")
     endif (nounused_avail)
-    # XXX i#3793: DynamoRIO manages '\0' termination and error states itself in
+    # XXX i#3792: DynamoRIO manages '\0' termination and error states itself in
     # too many places. In order to activate this warning, this code needs to get
     # re-factored for no good reason.
     CHECK_C_COMPILER_FLAG("-Wstringop-truncation" stringop_truncation_avail)
     if (stringop_truncation_avail)
       set(WARN "${WARN} -Wno-stringop-truncation")
     endif (stringop_truncation_avail)
-    # XXX i#3793: see comment above.
+    # XXX i#3792: see comment above.
     CHECK_C_COMPILER_FLAG("-Wformat-truncation" format_truncation_avail)
     if (format_truncation_avail)
       set(WARN "${WARN} -Wno-format-truncation")
     endif (format_truncation_avail)
-    # XXX i#3793: see comment above.
+    # XXX i#3792: see comment above.
     CHECK_C_COMPILER_FLAG("-Wstringop-overflow" stringop_overflow_avail)
     if (stringop_overflow_avail)
       set(WARN "${WARN} -Wno-stringop-overflow")

--- a/core/string.c
+++ b/core/string.c
@@ -117,32 +117,36 @@ d_r_strrchr(const char *str, int c)
 /* Private strncpy.  Standard caveat about not copying trailing null byte on
  * truncation applies.
  */
-#define D_R_STRNCPY_BODY                      \
-    size_t i;                                 \
-    for (i = 0; i < n && src[i] != '\0'; i++) \
-        dst[i] = src[i];                      \
-    /* Pad the rest with nulls. */            \
-    for (; i < n; i++)                        \
-        dst[i] = '\0';                        \
-    return dst;
+#define D_R_STRNCPY_BODY()                        \
+    do {                                          \
+        size_t i;                                 \
+        for (i = 0; i < n && src[i] != '\0'; i++) \
+            dst[i] = src[i];                      \
+        /* Pad the rest with nulls. */            \
+        for (; i < n; i++)                        \
+            dst[i] = '\0';                        \
+        return dst;                               \
+    } while (0);
 char *
 d_r_strncpy(char *dst, const char *src, size_t n)
 {
-    D_R_STRNCPY_BODY
+    D_R_STRNCPY_BODY()
 }
 
 /* Private strncat. */
-#define D_R_STRNCAT_BODY                      \
-    size_t dest_len = strlen(dest);           \
-    size_t i;                                 \
-    for (i = 0; i < n && src[i] != '\0'; i++) \
-        dest[dest_len + i] = src[i];          \
-    dest[dest_len + i] = '\0';                \
-    return dest;
+#define D_R_STRNCAT_BODY()                        \
+    do {                                          \
+        size_t dest_len = strlen(dest);           \
+        size_t i;                                 \
+        for (i = 0; i < n && src[i] != '\0'; i++) \
+            dest[dest_len + i] = src[i];          \
+        dest[dest_len + i] = '\0';                \
+        return dest;                              \
+    } while (0);
 char *
 d_r_strncat(char *dest, const char *src, size_t n)
 {
-    D_R_STRNCAT_BODY
+    D_R_STRNCAT_BODY()
 }
 
 /* Private memcpy is in arch/<arch>/<arch>.asm or memfuncs.asm */
@@ -156,21 +160,23 @@ d_r_strncat(char *dest, const char *src, size_t n)
  * DR libc isolation.
  */
 
-#define D_R_MEMMOVE_BODY                            \
-    ssize_t i;                                      \
-    byte *dst_b = (byte *)dst;                      \
-    const byte *src_b = (const byte *)src;          \
-    if (dst < src)                                  \
-        return memcpy(dst, src, n);                 \
-    /* FIXME: Could use reverse DF and rep movs. */ \
-    for (i = n - 1; i >= 0; i--) {                  \
-        dst_b[i] = src_b[i];                        \
-    }                                               \
-    return dst;
+#define D_R_MEMMOVE_BODY()                              \
+    do {                                                \
+        ssize_t i;                                      \
+        byte *dst_b = (byte *)dst;                      \
+        const byte *src_b = (const byte *)src;          \
+        if (dst < src)                                  \
+            return memcpy(dst, src, n);                 \
+        /* FIXME: Could use reverse DF and rep movs. */ \
+        for (i = n - 1; i >= 0; i--) {                  \
+            dst_b[i] = src_b[i];                        \
+        }                                               \
+        return dst;                                     \
+    } while (0);
 void *
 d_r_memmove(void *dst, const void *src, size_t n)
 {
-    D_R_MEMMOVE_BODY
+    D_R_MEMMOVE_BODY()
 }
 
 #ifdef UNIX
@@ -184,17 +190,17 @@ d_r_memmove(void *dst, const void *src, size_t n)
 void *
 __memmove_chk(void *dst, const void *src, size_t n, size_t dst_len)
 {
-    D_R_MEMMOVE_BODY
+    D_R_MEMMOVE_BODY()
 }
 void *
 __strncpy_chk(char *dst, const char *src, size_t n, size_t dst_len)
 {
-    D_R_STRNCPY_BODY
+    D_R_STRNCPY_BODY()
 }
 void *
 __strncat_chk(char *dest, const char *src, size_t n, size_t dst_len)
 {
-    D_R_STRNCAT_BODY
+    D_R_STRNCAT_BODY()
 }
 #endif
 

--- a/core/string.c
+++ b/core/string.c
@@ -117,28 +117,32 @@ d_r_strrchr(const char *str, int c)
 /* Private strncpy.  Standard caveat about not copying trailing null byte on
  * truncation applies.
  */
+#define D_R_STRNCPY_BODY                      \
+    size_t i;                                 \
+    for (i = 0; i < n && src[i] != '\0'; i++) \
+        dst[i] = src[i];                      \
+    /* Pad the rest with nulls. */            \
+    for (; i < n; i++)                        \
+        dst[i] = '\0';                        \
+    return dst;
 char *
 d_r_strncpy(char *dst, const char *src, size_t n)
 {
-    size_t i;
-    for (i = 0; i < n && src[i] != '\0'; i++)
-        dst[i] = src[i];
-    /* Pad the rest with nulls. */
-    for (; i < n; i++)
-        dst[i] = '\0';
-    return dst;
+    D_R_STRNCPY_BODY
 }
 
 /* Private strncat. */
+#define D_R_STRNCAT_BODY                      \
+    size_t dest_len = strlen(dest);           \
+    size_t i;                                 \
+    for (i = 0; i < n && src[i] != '\0'; i++) \
+        dest[dest_len + i] = src[i];          \
+    dest[dest_len + i] = '\0';                \
+    return dest;
 char *
 d_r_strncat(char *dest, const char *src, size_t n)
 {
-    size_t dest_len = strlen(dest);
-    size_t i;
-    for (i = 0; i < n && src[i] != '\0'; i++)
-        dest[dest_len + i] = src[i];
-    dest[dest_len + i] = '\0';
-    return dest;
+    D_R_STRNCAT_BODY
 }
 
 /* Private memcpy is in arch/<arch>/<arch>.asm or memfuncs.asm */
@@ -151,19 +155,22 @@ d_r_strncat(char *dest, const char *src, size_t n)
  * We also have a version named "memmove" in lib/memmove.c for shared
  * DR libc isolation.
  */
+
+#define D_R_MEMMOVE_BODY                            \
+    ssize_t i;                                      \
+    byte *dst_b = (byte *)dst;                      \
+    const byte *src_b = (const byte *)src;          \
+    if (dst < src)                                  \
+        return memcpy(dst, src, n);                 \
+    /* FIXME: Could use reverse DF and rep movs. */ \
+    for (i = n - 1; i >= 0; i--) {                  \
+        dst_b[i] = src_b[i];                        \
+    }                                               \
+    return dst;
 void *
 d_r_memmove(void *dst, const void *src, size_t n)
 {
-    ssize_t i;
-    byte *dst_b = (byte *)dst;
-    const byte *src_b = (const byte *)src;
-    if (dst < src)
-        return memcpy(dst, src, n);
-    /* FIXME: Could use reverse DF and rep movs. */
-    for (i = n - 1; i >= 0; i--) {
-        dst_b[i] = src_b[i];
-    }
-    return dst;
+    D_R_MEMMOVE_BODY
 }
 
 #ifdef UNIX
@@ -174,40 +181,19 @@ d_r_memmove(void *dst, const void *src, size_t n)
  */
 void *
 __memmove_chk(void *dst, const void *src, size_t n, size_t dst_len)
-#    ifdef MACOS
-/* OSX 10.7 gcc 4.2.1 doesn't support the alias attribute.
- * XXX: better to test for support at config time: for now assuming none on Mac.
- */
 {
-    return memmove(dst, src, n);
+    D_R_MEMMOVE_BODY
 }
-#    else
-    __attribute__((alias("d_r_memmove")));
-#    endif
 void *
 __strncpy_chk(char *dst, const char *src, size_t n, size_t dst_len)
-#    ifdef MACOS
-/* OSX 10.7 gcc 4.2.1 doesn't support the alias attribute.
- * XXX: better to test for support at config time: for now assuming none on Mac.
- */
 {
-    return strncpy(dst, src, n);
+    D_R_STRNCPY_BODY
 }
-#    else
-    __attribute__((alias("d_r_strncpy")));
-#    endif
 void *
-__strncat_chk(char *dst, const char *src, size_t n, size_t dst_len)
-#    ifdef MACOS
-/* OSX 10.7 gcc 4.2.1 doesn't support the alias attribute.
- * XXX: better to test for support at config time: for now assuming none on Mac.
- */
+__strncat_chk(char *dest, const char *src, size_t n, size_t dst_len)
 {
-    return strncat(dst, src, n);
+    D_R_STRNCAT_BODY
 }
-#    else
-    __attribute__((alias("d_r_strncat")));
-#    endif
 #endif
 
 /* Private strcmp. */

--- a/core/string.c
+++ b/core/string.c
@@ -126,11 +126,11 @@ d_r_strrchr(const char *str, int c)
         for (; i < n; i++)                        \
             dst[i] = '\0';                        \
         return dst;                               \
-    } while (0);
+    } while (0)
 char *
 d_r_strncpy(char *dst, const char *src, size_t n)
 {
-    D_R_STRNCPY_BODY()
+    D_R_STRNCPY_BODY();
 }
 
 /* Private strncat. */
@@ -142,11 +142,11 @@ d_r_strncpy(char *dst, const char *src, size_t n)
             dest[dest_len + i] = src[i];          \
         dest[dest_len + i] = '\0';                \
         return dest;                              \
-    } while (0);
+    } while (0)
 char *
 d_r_strncat(char *dest, const char *src, size_t n)
 {
-    D_R_STRNCAT_BODY()
+    D_R_STRNCAT_BODY();
 }
 
 /* Private memcpy is in arch/<arch>/<arch>.asm or memfuncs.asm */
@@ -172,11 +172,11 @@ d_r_strncat(char *dest, const char *src, size_t n)
             dst_b[i] = src_b[i];                        \
         }                                               \
         return dst;                                     \
-    } while (0);
+    } while (0)
 void *
 d_r_memmove(void *dst, const void *src, size_t n)
 {
-    D_R_MEMMOVE_BODY()
+    D_R_MEMMOVE_BODY();
 }
 
 #ifdef UNIX
@@ -190,17 +190,17 @@ d_r_memmove(void *dst, const void *src, size_t n)
 void *
 __memmove_chk(void *dst, const void *src, size_t n, size_t dst_len)
 {
-    D_R_MEMMOVE_BODY()
+    D_R_MEMMOVE_BODY();
 }
 void *
 __strncpy_chk(char *dst, const char *src, size_t n, size_t dst_len)
 {
-    D_R_STRNCPY_BODY()
+    D_R_STRNCPY_BODY();
 }
 void *
 __strncat_chk(char *dest, const char *src, size_t n, size_t dst_len)
 {
-    D_R_STRNCAT_BODY()
+    D_R_STRNCAT_BODY();
 }
 #endif
 

--- a/core/string.c
+++ b/core/string.c
@@ -176,8 +176,10 @@ d_r_memmove(void *dst, const void *src, size_t n)
 #ifdef UNIX
 /* gcc emits calls to these *_chk variants in release builds when the size of
  * dst is known at compile time.  In C, the caller is responsible for cleaning
- * up arguments on the stack, so we alias these *_chk routines to the non-chk
- * routines and rely on the caller to clean up the extra dst_len arg.
+ * up arguments on the stack. We used to alias these *_chk routines to the non-chk
+ * routines. Current gcc versions don't accept aliases with a different function
+ * signature. Instead, we are now manually inlining the function body. We rely
+ * on the caller to clean up the extra dst_len arg.
  */
 void *
 __memmove_chk(void *dst, const void *src, size_t n, size_t dst_len)

